File: pipeline/frontend/src/api/modelLab.ts
/** Model-lab queries: datasets, training runs, the model registry and prediction. */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { del, download, get, patch, post, postForm } from './client'
import { keys } from './keys'

export interface BackendDataset {
  id: number
  project_id: number
  original_name: string
  dataset_family: string | null
  row_count: number
  col_count: number
  headers: string[]
  column_types: Record<string, string>
  column_mapping: Record<string, string>
  parse_status: 'pending' | 'ready' | 'error'
  parse_error: string | null
  file_hash: string | null
  sheet_name: string | null
  uploaded_at: string
  updated_at: string
}

export interface ModelResult {
  id: number
  training_run_id: number
  model_name: string
  model_family: string
  status: 'pending' | 'training' | 'completed' | 'failed' | 'skipped'
  skip_reason: string | null
  error_message: string | null
  metrics: Record<string, number | null>
  parameters: Record<string, unknown>
  feature_cols: string[]
  target_col: string | null
  mae: number | null
  rmse: number | null
  r_squared: number | null
  concordance_index: number | null
  has_artifact: boolean
  is_active: boolean
  created_at: string
  completed_at: string | null
  results?: Record<string, unknown> | null
}

export interface TrainingRunStatus {
  id: number
  project_id: number
  dataset_id: number
  dataset_name: string | null
  job_id: number | null
  dataset_family: string
  n_trajectories: number
  n_fitted: number
  status: 'queued' | 'running' | 'completed' | 'failed'
  error_message: string | null
  created_at: string
  completed_at: string | null
  model_results: ModelResult[]
  job: {
    status: string
    progress: number
    current_step: string
    error_message: string | null
  } | null
}

export interface PredictResult {
  model_name: string
  predicted_shelf_life_days: number
  ci_lo_days: number | null
  ci_hi_days: number | null
  required_shelf_life_days?: number | null
  success?: boolean | null
  p_success?: number | null
  expected_spoilage_date?: string | null
}

const base = (projectId: number) => `/projects/${projectId}/model-lab`

// ─── Imperative calls ─────────────────────────────────────────────────────────
// The model-lab screens drive an explicit upload -> map -> train wizard where each step is
// a user action with its own progress UI, so they call these directly rather than through
// a query. They still go through the shared client, so auth, the base URL and the error
// envelope are identical; only the caching layer differs.

export const listDatasets = (projectId: number): Promise<BackendDataset[]> =>
  get<BackendDataset[]>(`${base(projectId)}/datasets`)

export const uploadDataset = async (
  projectId: number, file: File, family: string, forceReplace = false,
): Promise<{ duplicate: boolean; dataset: BackendDataset; message?: string }> => {
  const form = new FormData()
  form.append('file', file)
  form.append('dataset_family', family)
  form.append('force_replace', String(forceReplace))
  return postForm(`${base(projectId)}/datasets/upload`, form)
}

export const updateMapping = (
  projectId: number, datasetId: number,
  columnMapping: Record<string, string>, datasetFamily?: string,
): Promise<BackendDataset> =>
  patch<BackendDataset>(`${base(projectId)}/datasets/${datasetId}`, {
    column_mapping: columnMapping,
    dataset_family: datasetFamily,
  })

export const deleteDataset = (projectId: number, datasetId: number): Promise<void> =>
  del(`${base(projectId)}/datasets/${datasetId}`)

export const startTraining = (
  projectId: number, datasetId: number,
  columnMapping: Record<string, string>, datasetFamily: string, threshold?: number,
): Promise<{ training_run_id: number; job_id: number; status: string }> =>
  post(`${base(projectId)}/train`, {
    dataset_id: datasetId,
    column_mapping: columnMapping,
    dataset_family: datasetFamily,
    threshold,
  })

export const listRuns = (projectId: number): Promise<TrainingRunStatus[]> =>
  get<TrainingRunStatus[]>(`${base(projectId)}/runs`)

export const getRun = (projectId: number, runId: number): Promise<TrainingRunStatus> =>
  get<TrainingRunStatus>(`${base(projectId)}/runs/${runId}`)

export const listModels = (projectId: number): Promise<ModelResult[]> =>
  get<ModelResult[]>(`${base(projectId)}/models`)

export const runPrediction = (
  projectId: number, modelId: number, inputFeatures: Record<string, unknown>,
  requiredShelfLife?: number, startDate?: string,
): Promise<PredictResult> =>
  post<PredictResult>(`${base(projectId)}/predict`, {
    model_id: modelId,
    input_features: inputFeatures,
    required_shelf_life: requiredShelfLife,
    start_date: startDate,
  })

// ─── Datasets ─────────────────────────────────────────────────────────────────

export function useDatasets(projectId: number) {
  return useQuery({
    queryKey: keys.modelLab.datasets(projectId),
    queryFn: () => get<BackendDataset[]>(`${base(projectId)}/datasets`),
    enabled: Number.isFinite(projectId),
  })
}

export function useUploadDataset(projectId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ file, family, forceReplace = false }: {
      file: File
      family: string
      forceReplace?: boolean
    }) => {
      const form = new FormData()
      form.append('file', file)
      form.append('dataset_family', family)
      form.append('force_replace', String(forceReplace))
      return postForm<{ duplicate: boolean; dataset: BackendDataset; message?: string }>(
        `${base(projectId)}/datasets/upload`,
        form,
      )
    },
    onSuccess: () =>
      queryClient.invalidateQueries({ queryKey: keys.modelLab.datasets(projectId) }),
  })
}

export function useUpdateMapping(projectId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ datasetId, columnMapping, datasetFamily }: {
      datasetId: number
      columnMapping: Record<string, string>
      datasetFamily?: string
    }) =>
      patch<BackendDataset>(`${base(projectId)}/datasets/${datasetId}`, {
        column_mapping: columnMapping,
        dataset_family: datasetFamily,
      }),
    onSuccess: () =>
      queryClient.invalidateQueries({ queryKey: keys.modelLab.datasets(projectId) }),
  })
}

export function useDeleteDataset(projectId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (datasetId: number) => del(`${base(projectId)}/datasets/${datasetId}`),
    onSuccess: () =>
      queryClient.invalidateQueries({ queryKey: keys.modelLab.datasets(projectId) }),
  })
}

/**
 * Download a dataset through the axios client so it carries the bearer token and honours
 * the configured API base URL. The previous helper returned a bare `/api/...` string,
 * which bypassed `VITE_API_URL` entirely and broke whenever the gateway was not
 * same-origin.
 */
export const downloadDataset = (projectId: number, datasetId: number, name: string) =>
  download(`${base(projectId)}/datasets/${datasetId}/download`, name)

// ─── Training ─────────────────────────────────────────────────────────────────

export function useTrainingRuns(projectId: number) {
  return useQuery({
    queryKey: keys.modelLab.runs(projectId),
    queryFn: () => get<TrainingRunStatus[]>(`${base(projectId)}/runs`),
    enabled: Number.isFinite(projectId),
  })
}

export function useTrainingRun(projectId: number, runId: number | null) {
  return useQuery({
    queryKey: keys.modelLab.run(projectId, runId ?? 0),
    queryFn: () => get<TrainingRunStatus>(`${base(projectId)}/runs/${runId}`),
    enabled: runId !== null,
  })
}

export function useStartTraining(projectId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (body: {
      dataset_id: number
      column_mapping: Record<string, string>
      dataset_family: string
      threshold?: number
    }) =>
      post<{ training_run_id: number; job_id: number; status: string }>(
        `${base(projectId)}/train`,
        body,
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.modelLab.runs(projectId) })
      queryClient.invalidateQueries({ queryKey: keys.jobs.all() })
    },
  })
}

// ─── Registry and prediction ──────────────────────────────────────────────────

export function useModels(projectId: number) {
  return useQuery({
    queryKey: keys.modelLab.models(projectId),
    queryFn: () => get<ModelResult[]>(`${base(projectId)}/models`),
    enabled: Number.isFinite(projectId),
  })
}

export function useModel(projectId: number, modelId: number | null) {
  return useQuery({
    queryKey: keys.modelLab.model(projectId, modelId ?? 0),
    queryFn: () => get<ModelResult>(`${base(projectId)}/models/${modelId}`),
    enabled: modelId !== null,
  })
}

export function useDeleteModel(projectId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (modelId: number) => del(`${base(projectId)}/models/${modelId}`),
    onSuccess: () =>
      queryClient.invalidateQueries({ queryKey: keys.modelLab.models(projectId) }),
  })
}

export function usePredict(projectId: number) {
  return useMutation({
    mutationFn: (body: {
      model_id: number
      input_features: Record<string, unknown>
      required_shelf_life?: number
      start_date?: string
    }) => post<PredictResult>(`${base(projectId)}/predict`, body),
  })
}

File: pipeline/processing/app/schemas/model_lab.py
"""Model-lab request and response models.

Replaces the hand-built dicts (`_dataset_to_dict`, `_run_to_dict`, `_result_to_dict`) that
lived in the route module and were the de-facto API contract without ever being declared.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class DatasetOut(BaseModel):
    id: int
    project_id: int
    original_name: str
    dataset_family: Optional[str] = None
    row_count: int = 0
    col_count: int = 0
    headers: list[str] = []
    column_types: dict[str, str] = {}
    column_mapping: dict[str, str] = {}
    parse_status: str
    parse_error: Optional[str] = None
    file_hash: Optional[str] = None
    sheet_name: Optional[str] = None
    uploaded_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class DatasetUploadOut(BaseModel):
    duplicate: bool
    dataset: DatasetOut
    message: Optional[str] = None


class MappingUpdate(BaseModel):
    column_mapping: dict[str, str]
    dataset_family: Optional[str] = None


class ModelResultOut(BaseModel):
    id: int
    training_run_id: int
    model_name: str
    model_family: str
    status: str
    skip_reason: Optional[str] = None
    error_message: Optional[str] = None
    metrics: dict[str, Any] = {}
    parameters: dict[str, Any] = {}
    feature_cols: list[str] = []
    target_col: Optional[str] = None
    mae: Optional[float] = None
    rmse: Optional[float] = None
    r_squared: Optional[float] = None
    concordance_index: Optional[float] = None
    has_artifact: bool = False
    is_active: bool = True
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    #: Full per-trajectory fits; only populated on the single-model endpoint for kinetic
    #: families, because it can be megabytes and would bloat every list response.
    results: Optional[dict[str, Any]] = None


class JobSummary(BaseModel):
    status: str
    progress: int = 0
    current_step: str = ""
    error_message: Optional[str] = None


class TrainingRunOut(BaseModel):
    id: int
    project_id: int
    dataset_id: int
    dataset_name: Optional[str] = None
    job_id: Optional[int] = None
    dataset_family: str
    n_trajectories: int = 0
    n_fitted: int = 0
    status: str
    error_message: Optional[str] = None
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    model_results: list[ModelResultOut] = []
    job: Optional[JobSummary] = None


class TrainRequest(BaseModel):
    dataset_id: int
    column_mapping: dict[str, str]
    dataset_family: str
    #: log10 threshold defining failure for kinetic time-to-failure derivation.
    threshold: Optional[float] = None


class TrainingAccepted(BaseModel):
    training_run_id: int
    job_id: int
    status: str


class PredictRequest(BaseModel):
    model_id: int
    input_features: dict[str, Any] = Field(default_factory=dict)
    required_shelf_life: Optional[float] = None
    #: ISO date; when given, the response carries the expected spoilage date.
    start_date: Optional[str] = None


class PredictionOut(BaseModel):
    model_name: str
    predicted_shelf_life_days: float
    ci_lo_days: Optional[float] = None
    ci_hi_days: Optional[float] = None
    required_shelf_life_days: Optional[float] = None
    success: Optional[bool] = None
    p_success: Optional[float] = None
    expected_spoilage_date: Optional[str] = None
